### DISCUSSION TITLE
Different results between netwrokx and numpy to count free balls

### DISCUSSION THREAD

**Discussion by alisheikholeslam** [networkx/networkx#5963] [Q&A] [answered]
There are some balls that are almost connected to each other.

[![enter image description here][1]][1]

I have written two different codes (one by [*numpy and scipy cKDTree*](https://stackoverflow.com/q/71104627/13394817) and one by [*networkx*](https://stackoverflow.com/q/72537736/13394817) to find number of balls that are free (with no connection):
```
import numpy as np
from scipy.spatial import cKDTree, distance
import networkx as nx


radii = np.load('radii_freeballs.npy')
poss = np.load('poss_freeballs.npy')

def ends_gap_numpy_opt(poss, radii):

    rad_max = radii.max()
    particle_corsp_overlaps = []
    ends_ind = []
    kdtree = cKDTree(poss)

    for particle_idx in range(len(poss)):

        cur_point = poss[particle_idx]
        dist_max = radii[particle_idx] + rad_max
        nears_i_ind = np.array(kdtree.query_ball_point(cur_point, r=dist_max, return_sorted=True), dtype=np.int64)
        assert len(nears_i_ind) > 0

        if len(nears_i_ind) <= 1:
            continue

        nears_i_ind = nears_i_ind[nears_i_ind != particle_idx]
        dist_i = distance.cdist(poss[nears_i_ind], cur_point[None, :]).squeeze()

        contact_check = dist_i - (radii[nears_i_ind] + radii[particle_idx])
        connected = contact_check[contact_check <= 0]
        particle_corsp_overlaps.append(connected)

        contacts_ind = np.where([contact_check <= 0])[1]
        contacts_sec_ind = nears_i_ind[contacts_ind]
        sphere_olps_ind = np.sort(contacts_sec_ind)

        ends_ind_mod_temp = np.array([np.repeat(particle_idx, len(sphere_olps_ind)), sphere_olps_ind], dtype=np.int64).T
        ends_ind.append(ends_ind_mod_temp)

    ends_ind_org = np.concatenate(ends_ind)
    ends_ind, ends_ind_idx = np.unique(np.sort(ends_ind_org), axis=0, return_index=True)

    return ends_ind


# using networkx
maximum_distances = radii.reshape(radii.shape[0], 1) + radii.reshape(1, radii.shape[0])
distances = distance.cdist(poss, poss)
connections = distances < maximum_distances
G = nx.Graph(connections)
connected_paths = list(nx.connected_components(G))
paths_lens = np.array([len(c) for c in connected_paths])
print("free_balls networkx", (paths_lens == 1).sum())

# using numpy & scipy cKDTree
ends_ind = ends_gap_numpy_opt(poss, radii)
contact_counts = np.pad(np.bincount(np.ravel(ends_ind)), [(0, len(poss) - len(np.bincount(np.ravel(ends_ind))))], mode='constant')
print("free_balls numpy", (contact_counts == 0).sum())
```
**I don't know what is the reason for different results of these two methods, and how to fix it.** E.g., for the prepared example, networkx count 17 free balls but the other method count 4; As it can be seen from the pic, the *numpy & scipy cKDTree* solution is consistent with the number of free balls that are determined by a simulator in the pic (it is checked with different data).  
Note that, the `ends_gap_numpy_opt` function, which shows indices of each two connected spheres, is previously checked by other proposed functions and the simulator result in terms of correctness.

**Data:**  
Radii: https://drive.google.com/file/d/1epnIoIh9z9w0w5leQ1zP2JVgsouBFxFM/view?usp=sharing  
Poss : https://drive.google.com/file/d/1SOwfsrM_sE77tOrRkIUvX5AyFA-0FHCo/view?usp=sharing

  [1]: https://i.stack.imgur.com/iknVU.png

**Comment by dschult** [comment, marked as answer]
This question is pretty complex to dive into -- can you tell somehow that the networks that result from the two approaches are the same?  That would be a first step. It's hard for me to see from this code whether the two methods should even give the same result. One possible difference is the < in the networkx method may differ from <= in the contact _check. But there are so many things going on that there could be lots of other subtle differences in methods.

**Reply by alisheikholeslam** [reply]
Thanks, I found the problem in my code. Just to know, which float types networkx using for calculation? Are they float64?
